core/templates/dev/head/components/forms/FormBuilderSpec.js

```diff
@@ -146,6 +169,212 @@ describe('Normalizer tests', function() {
+    element = '<form name="testForm">' +
+      '<input name="inputValue" type="number" ng-model="localValue" ' +
+      'apply-validation>' +
+      '</form>';
+  }));
+
+  it('should apply isAtLeast validation', inject(function($compile) {
+    scope.validators = function() {
+      return [{
+        id: 'isAtLeast',
+        minValue: -2.5
+      }];
+    };
+    $compile(element)(scope);
+    testInput = scope.testForm.inputValue;
+
+    testInput.$setViewValue(-1);
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue('1');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue(-3);
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+
+    testInput.$setViewValue('-3');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+  }));
+
+  it('should apply isAtMost validation', inject(function($compile) {
+    scope.validators = function() {
+      return [{
+        id: 'isAtMost',
+        maxValue: 5
+      }];
+    };
+    $compile(element)(scope);
+    testInput = scope.testForm.inputValue;
+
+    testInput.$setViewValue(-1);
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue('1');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue(6);
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+
+    testInput.$setViewValue('10');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+  }));
+
+  it('should apply isNonempty validation', inject(function($compile) {
+    scope.validators = function() {
+      return [{
+        id: 'isNonempty'
+      }];
+    };
+    $compile(element)(scope);
+    testInput = scope.testForm.inputValue;
+
+    testInput.$setViewValue(-1);
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue('1');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue('');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+
+    testInput.$setViewValue(undefined);
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+  }));
+
+  it('should apply isInteger validation', inject(function($compile) {
+    scope.validators = function() {
+      return [{
+        id: 'isInteger'
+      }];
+    };
+    $compile(element)(scope);
+    testInput = scope.testForm.inputValue;
+
+    testInput.$setViewValue(-3);
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue('1');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue('3');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue(3.5);
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+  }));
+
+  it('should apply isFloat validation', inject(function($compile) {
```

Comment:
Interesting. Is the result ever false? Maybe add a test that uses ints?

Reply:
Hm, I took a look at the original method and I can't see how it returns false either. It seems a bit weird actually -- whereas all the other isInteger, isNonempty, etc. methods return true or false, this one seems to return a float or undefined. That seems inconsistent. But it's great we're catching this -- that's what happens when you write tests :)

I suggest changing the main code to return true/false instead, and updating the tests accordingly. Just make sure to take a look at all instances of isFloat  -- e.g. in SchemaBasedFloatEditorDirective.js -- and verify that everything still works correctly. Also test manually with a float editor (e.g. in the NumericInput interaction perhaps?) to see whether everything still behaves OK when you type in valid input and when you do not.

If there are issues, then maybe what's really needed is two filters -- isFloat, and parseAsFloat. But let's see if we can do without the second one.

Does that sound good?